python_modules/dagster-externals/dagster_externals/_io/base.py

```diff
@@ -0,0 +1,46 @@
+from abc import ABC, abstractmethod
+
+from typing_extensions import Self
+
+from .._protocol import (
+    ExternalExecutionContextData,
+    ExternalExecutionMessage,
+)
+from .._util import (
+    emit_orchestration_inactive_warning,
+    get_mock,
+    is_dagster_orchestration_active,
+    param_from_env,
+)
+
+
+class ExternalExecutionContextSource(ABC):
+    @classmethod
+    def from_env(cls) -> Self:
+        if not is_dagster_orchestration_active():
+            emit_orchestration_inactive_warning()
+            return get_mock()
+        else:
+            context_source_params = param_from_env("context_source")
+            assert isinstance(context_source_params, dict)
+            return cls(**context_source_params)
```

Comment:
seems like we should pick one spot to do mocks. Right now it is both here and `init_dagster_externals`. Recommend _only_ doing it in `init_dagster_externals`

Reply:
Can you elaborate-- based on discussion I thought we wanted to have users directly instantiate context source/message sink objects. Since they are being passed into `init_dagster_external`, they are being instantiated before it is called. If they are reading their params from the env (or wherever the orchestration process injected them), then those aren't going to be available if orchestration isn't active. So to only do mocks in one spot, I see two options, neither of which are particularly appealing:

- Make sure context source / message sink can tolerate null params
- Require an additional `setup()` method or something after instantiation for these objects, which reads the params from the environment and mocks at this time. This would be called in `init_dagster_externals`.

Reply:
Yeah lack of nice, first-class lambdas in python is so painful. To me the right solution here for these to be lambdas and then create the source/sink on demand after the env check.

While we are at it I'm not a huge of of the placement and naming `from_env` since it is implies it will work for all launch strategies and context source types, which it will not. It also implies an interface on the `__init__` of the subclass which is not enforcable. (Any pattern like this is better served by a abstract class method with a well-defined interface).

My recommendation is to change this so that all environment access stuff is done on demand in `load_context`.

Reply:
I moved the env logic out of the ABC and into the default file-based implementations.